Modules/mathmodule.c

```diff
@@ -2561,8 +2561,8 @@ math_prod_impl(PyObject *module, PyObject *iterable, PyObject *start)
                 long x = i_result * b;
                 /* Continue if there is no overflow */
                 if (overflow == 0
-                    && x < LONG_MAX && x > LONG_MIN
-                    && !(b != 0 && x / b != i_result)) {
+                    && !(b != 0 && (i_result > LONG_MAX / b
```

Comment:
Does this code work in the case that `b` is negative?

Reply:
Hummm.... Unless I am missing something, it does:

```python
import math, functools, operator
x = [1,2,3,4,-1 * (1<<1000),3,4,5,6,7,]
>>> functools.reduce(operator.mul, x) == math.prod(x)
True
```

Reply:
Let me rephrase: this code _doesn't_ work in the case that `b` is negative. :-)  The sense of the comparisons needs to be reversed in that case.

From another comment:

> @mdickinson Do you have some suggestions for extra checks to cover these corner cases?

I'd start out by declaring `LONG_MIN` an overflow case (even though it isn't), and assume that `LONG_MIN + LONG_MAX = -1` (which isn't guaranteed by C, but is overwhelmingly likely to be true on any hardware that Python cares about; the assumption is probably worth a comment in the source, but it's not worth worrying about beyond that). Then we just have to keep track of whether we've overflowed `LONG_MAX` in absolute value, and that's much simpler to do; it also takes care of the dividing `LONG_MIN` by `-1` problem. (This is where we win over Python 2's int * int: in that case, it's important not to treat a result of `LONG_MIN` as an overflow case, but here we don't care.)

But re-using the Python 2 overflow detection works, too.

Reply:
@mdickinson Thanks for the explanation!

I had changed the code to use Python2's handling of the overflow because I prefer to go with battle-tested checks given how tricky multiplication overflow is. In the future, we could relax the test to use `LONG_MAX` and `LONG_MIN` if you think that is ok :)